### PR TITLE
Squashed commit of NVIDIA changes for CUDA9/Volta

### PR DESCRIPTION
**Pull request by borisfom**
This may not be a PR per se - more like work in progress.

- I have implemented tensor_op setting for Volta which seems to work. This should be extended to try fp16 HMMA - in the next cycle.

A larger chunk of changes is about algo cache. I am happy with the refactoring part - it works. Actual caching is disabled as It still needs work, I am looking for your help:
1. Shape string generation needs to be fixed
2. Alignment part of the hash key should be added (from C code?)
3. device id needs to be added to the hash key, as well as fwd/bwd_filter/bwd_data selector.

### CODE REVIEW

File: theano/configdefaults.py [hunk 1]
```diff
@@ -1580,7 +1580,7 @@ def filter_vm_lazy(val):
 
 
 def _timeout_default():
-    return theano.config.compile.wait * 24
+    return theano.config.compile.wait * 48
```
Comment (nouiz) [inline theano/configdefaults.py]: why change this? The current number is 5*24, so 120s. What prompted you to change this?

File: theano/gpuarray/__init__.py [hunk 2]
```diff
@@ -123,7 +123,8 @@ def init_dev(dev, name=None, preallocate=None):
               (name, dev, context.devname, pcibusid),
               file=sys.stderr)
     pygpu_activated = True
-
+    cuda_available = True
```
Comment (nouiz) [inline theano/gpuarray/__init__.py]: We don't use that variable anymore. This was the name in the old back-end. User should start to use pygpu_activated instead.

If this is difficult for user, can you explain why? To have this change work well, it would request other change to this file.

File: theano/gpuarray/cudnn_defs.py [hunk 3]
```diff
@@ -145,7 +151,7 @@ def get_definitions(cudnn_version=None):
     if None, return definitions for the  most recent supported cuDNN version.
 
     """
-    if cudnn_version is not None and cudnn_version // 1000 == 5:
-        return CuDNNV51()
+    if cudnn_version is not None and cudnn_version // 1000 == 6:
```
Comment (nouiz) [inline theano/gpuarray/cudnn_defs.py]: You seem to keep cudnn v5 support. I'm fine to keep it or remove it now. If it is easy to keep, why not.

If we keep it, you would need to add this:

elif cudnn_version is not None and cudnn_version // 1000 == 6:
    return CuDNNV51()

File: theano/gpuarray/dnn.py [hunk 4]
```diff
@@ -58,7 +58,7 @@
     pass
 
 # Update these names when new versions of cudnn are supported.
-WIN32_CUDNN_NAMES = ['cudnn64_6.dll', 'cudnn64_5.dll']
+WIN32_CUDNN_NAMES = ['cudnn64_7.dll', 'cudnn64_6.dll', 'cudnn64_5.dll']
```
Comment (nouiz) [inline theano/gpuarray/dnn.py]: If we remove v5 support, remove the last element of this list.

File: theano/gpuarray/dnn.py [hunk 5]
```diff
@@ -164,13 +164,13 @@ def _dnn_check_compile():
 
 def _dnn_check_version():
     v = version()
-    if v < 5000:
-        return False, "cuDNN version is too old. Update to v5* or higher, was %d." % v
-    if v >= 6100:
+    if v < 6000:
+        return False, "cuDNN version is too old. Update to v6* or higher, was %d." % v
+    if v >= 7200:
```
Comment (nouiz) [inline theano/gpuarray/dnn.py]: why 7200? I would keep doing as before and use 7100. This way user will know we didn't tested that version of Theano with a new release.

File: theano/gpuarray/dnn_fwd.c [hunk 6]
```diff
@@ -142,6 +143,9 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
           return 1;
         }
         algo = choice.algo;
+        params->conv_algo = algo;
+        params->conv_ws_size = choice.memory;
+        params->conv_tensor_op = choice.mathType;
```
Comment (nouiz) [inline theano/gpuarray/dnn_fwd.c]: I don't think this is a good use of params. Can you just create new variable like algo (around line 86) instead for those 3 extra information?

Why did you needed to modify params?

File: theano/gpuarray/dnn_fwd.c [hunk 7]
```diff
@@ -172,7 +176,18 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
     } else {
       algo = prev_algo;
     }
-
+    
+    params->conv_algo = algo;
+    // CUDNN7: need to set math type
+    err = cudnnSetConvolutionMathType(desc, params->conv_tensor_op);
```
Comment (nouiz) [inline theano/gpuarray/dnn_fwd.c]: This need a "#ifdef CUDNN>=7" or similar I think. In this file, it is not like in libgpuarray, we include cudnn.h and I think this function don't exist in cudnn v6

File: theano/gpuarray/dnn_gw.c [hunk 8]
```diff
@@ -154,8 +154,8 @@ APPLY_SPECIFIC(conv_gw)(PyGpuArrayObject *input, PyGpuArrayObject *output,
         return 1;
       }
 
-      // Guess 4Mb if the info is not available
-      if (free == 0) free = 4 * 1024 * 1024;
+      // Guess 1Gb if the info is not available
+      if (free == 0) free = 1 * 1024 * 1024 * 1024;
```
Comment (nouiz) [inline theano/gpuarray/dnn_gw.c]: Why this change? I think it is risky to use an high value here.

File: theano/gpuarray/subtensor.py [hunk 9]
```diff
@@ -1108,15 +1107,19 @@ def gpu_kernels(self, node, nodename):
...
+      __half2float(old_perm));
     new_ = __byte_perm(old, sum, ((ga_size)addr & 2) ? 0x5410 : 0x3254);
     old = atomicCAS(base, assumed, new_);
   } while (assumed != old);
-  return (ga_half)__byte_perm(old, 0,
-                                  ((ga_size)addr & 2) ? 0x4432 : 0x4410);
+  ga_half ret;
+  __HALF_TO_US(ret) = __byte_perm(old, 0,
+        ((ga_size)addr & 2) ? 0x4432 : 0x4410);
+  return ret;
```
Comment (nouiz) [inline theano/gpuarray/subtensor.py]: @abergeron I let you review this part.

File: theano/scalar/basic.py [hunk 10]
```diff
@@ -2658,6 +2658,13 @@ class Neg(UnaryScalarOp):
     nfunc_spec = ('negative', 1, 1)
 
     def impl(self, x):
+        # We have to make sure x is not a numpy.bool_, because
+        # `-numpy.bool_(True)` is `False` (we want 0), and
```
Comment (nouiz) [inline theano/scalar/basic.py]: numpy.bool_ is python bool. numpy.bool is a numpy scalar of type bool. 

So mostly, the problem is with python bool, not numpy bool.

It would be great to add a test for that. Do you still have the problematic case? If not, I'll try to make one.

File: theano/gpuarray/dnn_fwd.c (hunk 6 above)
Comment (borisfom) [inline theano/gpuarray/dnn_fwd.c]: This is because I want to access new values from the caller in Python.

File: theano/gpuarray/dnn_gw.c (hunk 8 above)
Comment (borisfom) [inline theano/gpuarray/dnn_gw.c]: 4MB is very low for common nets - I have seen larger than 1G workspaces required.

File: theano/gpuarray/subtensor.py (hunk 9 above)
Comment (borisfom) [inline theano/gpuarray/subtensor.py]: Yes this is previously done CUDA9 fix.

File: theano/scalar/basic.py (hunk 10 above)
Comment (borisfom) [inline theano/scalar/basic.py]: Please do!

File: theano/gpuarray/dnn.py (hunk 5 above)
Comment (borisfom) [inline theano/gpuarray/dnn.py]: People were talking about 7.1 coming soon ... 7.1 is fine.

File: theano/gpuarray/cudnn_defs.py (hunk 3 above)
Comment (borisfom) [inline theano/gpuarray/cudnn_defs.py]: I meant to remove it - just did not see the point of modifying existing structs.

File: theano/gpuarray/__init__.py (hunk 2 above)
Comment (borisfom) [inline theano/gpuarray/__init__.py]: This is an artifact, please remove

File: theano/configdefaults.py (hunk 1 above)
Comment (borisfom) [inline theano/configdefaults.py]: I had a timeout once - we can revert it

File: requirement-rtd.txt [hunk 11]
```diff
@@ -1,6 +1,12 @@
 sphinx>=1.3.0
 pygments
 nose>=1.3.0
+numpy
+gnumpy
+pydot
+pydot2
+Cython
```
Comment (borisfom) [inline requirement-rtd.txt]: It sort of worked in our script - for Docker at least.

File: doc/index.txt [hunk 12]
```diff
@@ -20,6 +20,9 @@ learning/machine learning <https://mila.umontreal.ca/en/cours/>`_ classes).
 
 News
 ====
+* 2017/03/07: Release of Theano 0.9.0rc3, with crash fixes, bug fixes and improvements.
+
+* 2017/03/20: Release of Theano 0.9.0. Everybody is encouraged to update.
```
Comment (borisfom) [inline doc/index.txt]: Artifact

File: clean.sh [hunk 13]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
```
Comment (borisfom) [inline clean.sh]: Please remove - this is stuff I added for NV release

File: theano/gpuarray/dnn.py (hunk 5 above)
Comment (nouiz) [inline theano/gpuarray/dnn.py]: k

File: theano/gpuarray/subtensor.py (hunk 9 above)
Comment (abergeron) [inline theano/gpuarray/subtensor.py]: I want to move those definitions to libgpuarray so that we don't need to provide them manually everywhere.

File: theano/gpuarray/dnn_fwd.c (hunk 6 above)
Comment (abergeron) [inline theano/gpuarray/dnn_fwd.c]: Params doesn't allow you to do that.  It's one-way only.

File: theano/gpuarray/dnn_gw.c (hunk 8 above)
Comment (abergeron) [inline theano/gpuarray/dnn_gw.c]: This is fallback code in case we can't get the real amount of free memory.  We don't want to assume that there is a very large amount available and then fail to allocate it.

**Comment by borisfom** [comment]
Will reopen on a new branch.
